Add aggregation to common python timer.

LGM-Doyle opened the pull request:
This PR moves the portions of the AI timer that are not dependent on the `freeOrionAIInterface` into the python common code.  It then adds aggregation to the timer.  Aggregation means adding up the times of recursive sections of code and printing the sum with `print_aggregate()`.  

The original style of printing is available with `print_flat()`.  Calling `end()` still defaults to `print_flat()`.  

This addition to the `Timer` helped in profiling the universe generation code.

LGM-Doyle committed:
Move common portions to python AI Timer into common.

This refactors AI/freeorion_tools/timers and moves the portions that do
not depend in freeOrionAIInterface into python/common.

It renames the FO AI dependent timer to FOTimer.

LGM-Doyle committed:
PEP 8 fixes in common/timers.py

1. Removed brackets from print.
2. Moved declaration for self.end_time in __init__.

LGM-Doyle committed:
Add print_aggregate to python/common/timer to aggregate recursive timers.

This commit renames the default print function to print_flat().

It adds a second print function print_aggregate() that accumulates the
results from recursive timers and prints that sum as a single category.

LGM-Doyle committed:
For review #1223 rename FOTimer to AITimer.

As pointed out by Cjkjvfnby the name is more descriptive since it is the
timer dependent of the AI engine.

LGM-Doyle committed:
For review #1223 added period to end of comments.

LGM-Doyle committed:
For review #1223 add TODO to remind to use common/print_utils code.

LGM-Doyle committed:
Rename python Timers.clear() to clear_data()

This commit addresses concerns from review #1223 that clear() is ambiguous by:
- Renaming clear() to clear_data()
- Adding an explicit comment for the AI timers to clarify that clear_data()
  clears the timers each turn.
- Removing the unneeded clear functionality,  "self.timers=[]", from the
  AITimers.end() function.

LGM-Doyle committed:
Rename python Timer.end() to stop_and_print().

This clarifies ambiguity in the naming of end() raised in review #1223.

LGM-Doyle committed:
In Timer._print_timer_table Remove header/footer references to self.timers.

Cjkjvfnby reviewed (changes requested):
I see next timers commands:
 - `start(section)` start new section with name
 - `stop(additional_info="")` stops current section adds additional_info to its name
 - print(aggregate=False)  all prints goes here
 - reset() Reset timer to initial state 

But I don't see usecases where I `reset` timer and don't `print` its value.  It is place to mistake, forget to reset timer -> get growing print each time.  And I already see such place in your code.  **Any reason why print and reset should not be merged to single method?**

For aggregated add next columns to table:  number of calls and mean time.  At least TODO, because it should be done along with  adding Table support in the same time.

LGM-Doyle committed:
Add stop_print_and_clear() to the python AITimer.

This addresses issues from review #1223.

LGM-Doyle committed:
Add print_statistics() to python Timer.

print_statistics() computes and prints the number of samples, mean,
standard deviation and the maximum time measured.

LGM-Doyle committed:
Move python Timer.log_name initialization to AITimer where it is used.
